### PR TITLE
Stabilize `min_exhaustive_patterns`

### PR DESCRIPTION
**Pull request by Nadrieril**
## Stabilisation report

I propose we stabilize the [`min_exhaustive_patterns`](https://github.com/rust-lang/rust/issues/119612) language feature.

With this feature, patterns of empty types are considered unreachable when matched by-value. This allows:
```rust
enum Void {}
fn foo() -> Result<u32, Void>;

fn main() {
  let Ok(x) = foo();
  // also
  match foo() {
    Ok(x) => ...,
  }
}
```

This is a subset of the long-unstable [`exhaustive_patterns`](https://github.com/rust-lang/rust/issues/51085) feature. That feature is blocked because omitting empty patterns is tricky when *not* matched by-value. This PR stabilizes the by-value case, which is not tricky.

The not-by-value cases (behind references, pointers, and unions) stay as they are today, e.g.
```rust
enum Void {}
fn foo() -> Result<u32, &Void>;

fn main() {
  let Ok(x) = foo(); // ERROR: missing `Err(_)`
}
```

The consequence on existing code is some extra "unreachable pattern" warnings. This is fully backwards-compatible.

### Comparison with today's rust

This proposal only affects match checking of empty types (i.e. types with no valid values). Non-empty types behave the same with or without this feature. Note that everything below is phrased in terms of `match` but applies equallly to `if let` and other pattern-matching expressions.

To be precise, a visibly empty type is:
- an enum with no variants;
- the never type `!`;
- a struct with a *visible* field of a visibly empty type (and no #[non_exhaustive] annotation);
- a tuple where one of the types is visibly empty;
- en enum with all variants visibly empty (and no `#[non_exhaustive]` annotation);
- a `[T; N]` with `N != 0` and `T` visibly empty;
- all other types are nonempty.

(An extra change was proposed below: that we ignore #[non_exhaustive] for structs since adding fields cannot turn an empty struct into a non-empty one)

For normal types, exhaustiveness checking requires that we list all variants (or use a wildcard). For empty types it's more subtle: in some cases we require a `_` pattern even though there are no valid values that can match it. This is where the difference lies regarding this feature.

#### Today's rust

Under today's rust, a `_` is required for all empty types, except specifically: if the matched expression is of type `!` (the never type) or `EmptyEnum` (where `EmptyEnum` is an enum with no variants), then the `_` is not required.

```rust
let foo: Result<u32, !> = ...;
match foo {
    Ok(x) => ...,
    Err(_) => ..., // required
}
let foo: Result<u32, &!> = ...;
match foo {
    Ok(x) => ...,
    Err(_) => ..., // required
}
let foo: &! = ...;
match foo {
    _ => ..., // required
}
fn blah(foo: (u32, !)) {
    match foo {
        _ => ..., // required
    }
}
unsafe {
    let ptr: *const ! = ...;
    match *ptr {} // allowed
    let ptr: *const (u32, !) = ...;
    match *ptr {
        (x, _) => { ... } // required
    }
    let ptr: *const Result<u32, !> = ...;
    match *ptr {
        Ok(x) => { ... }
        Err(_) => { ... } // required
    }
}
```

#### After this PR

After this PR, a pattern of an empty type can be omitted if (and only if):
- the match scrutinee expression has type  `!` or `EmptyEnum` (like before);
- *or* the empty type is matched by value (that's the new behavior).

In all other cases, a `_` is required to match on an empty type.

```rust
let foo: Result<u32, !> = ...;
match foo {
    Ok(x) => ..., // `Err` not required
}
let foo: Result<u32, &!> = ...;
match foo {
    Ok(x) => ...,
    Err(_) => ..., // required because `!` is under a dereference
}
let foo: &! = ...;
match foo {
    _ => ..., // required because `!` is under a dereference
}
fn blah(foo: (u32, !)) {
    match foo {} // allowed
}
unsafe {
    let ptr: *const ! = ...;
    match *ptr {} // allowed
    let ptr: *const (u32, !) = ...;
    match *ptr {
        (x, _) => { ... } // required because the matched place is under a (pointer) dereference
    }
    let ptr: *const Result<u32, !> = ...;
    match *ptr {
        Ok(x) => { ... }
        Err(_) => { ... } // required because the matched place is under a (pointer) dereference
    }
}
```

### Documentation

The reference does not say anything specific about exhaustiveness checking, hence there is nothing to update there. The nomicon does, I opened https://github.com/rust-lang/nomicon/pull/445 to reflect the changes.

### Tests

The relevant tests are in `tests/ui/pattern/usefulness/empty-types.rs`.

### Unresolved Questions

None that I know of.

try-job: dist-aarch64-apple

### CODE REVIEW

**Comment by Nadrieril** [comment]
lol I did *not* intend to hit "send" now

**Comment by Nadrieril** [comment]
r? compiler for the normal PR review

**Comment by oxalica** [comment]
> * a struct with a _visible_ field of empty type (and no `#[non_exhaustive]` annotation);

Why `non_exhaustive` is excluded here? Unlike the `enum` case, if the struct already has a visible field of empty type, further added fields cannot make it non-empty again.

File: compiler/rustc_pattern_analysis/src/usefulness.rs
Comment (fee1-dead) [inline compiler/rustc_pattern_analysis/src/usefulness.rs:887]: why was empty_arms_are_unreachable replaced?

File: tests/ui/codemap_tests/huge_multispan_highlight.svg
Comment (fee1-dead) [inline tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: how did this change..?

File: tests/ui/pattern/usefulness/empty-types.normal.stderr
Comment (fee1-dead) [inline tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: not a blocker to stabilization, but these error messages can be a little bit confusing. We might want to add more context to it rather than a generic message.

**Review by fee1-dead — APPROVED** [review]
PR mostly looks fine.

File: tests/ui/codemap_tests/huge_multispan_highlight.svg
Comment (lqd) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: Maybe the PR just needs a rebase to get rid of this, as https://github.com/rust-lang/rust/pull/122779 removed some non determinism in these SVG snapshots.

File: compiler/rustc_pattern_analysis/src/usefulness.rs
Comment (Nadrieril) [reply compiler/rustc_pattern_analysis/src/usefulness.rs:887]: I duplicated because it felt like clear, that call is just a `matches!()` it's pretty much free

File: tests/ui/pattern/usefulness/empty-types.normal.stderr
Comment (Nadrieril) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: On the unreachable ones? I was thinking we should have a better message on the ones that still go "Err() required"

Comment (fee1-dead) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: When the unreachable pattern is caused by a type being uninhabited, we could point at it and say it is uninhabited. Oh and also it might be nice to add a suggestion for it to be removed?

**Comment by jplatte** [comment]
Related to what @oxalica said, an enum with no variants but the non-exhaustive attribute is surely not be considered empty for this? (first bullet point)

**Comment by Nadrieril** [comment]
Indeed this is not considered empty:
```rust
#[non_exhaustive]
enum Empty {}
```

For structs yeah you're right it would make sense to consider this empty:
```rust
#[non_exhaustive]
struct Struct {
  pub never: !,
}
```
I'm not sure why this is not the case today. Sounds sensible to change, I'll see what T-lang thinks.

File: tests/ui/pattern/usefulness/empty-types.normal.stderr
Comment (Nadrieril) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: > When the unreachable pattern is caused by a type being uninhabited, we could point at it and say it is uninhabited.

Good point, that would be helpful.

> Oh and also it might be nice to add a suggestion for it to be removed?

For what to be removed? The type?

Comment (fee1-dead) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: The match arm.

Comment (Nadrieril) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: Isn't that already what the "unreachable pattern" lint is saying?

Comment (Nadrieril) [reply tests/ui/pattern/usefulness/empty-types.normal.stderr:266]: Oh you mean a structured suggestion? eah that makes sense

**Comment by RalfJung** [comment]
Note that this feature *does* consider (some) structs with an uninhabited field as uninhabited, see the PR description. There seems to be a misunderstanding?

Also crucially, "by value" excludes the contents of unions.

**Comment by jplatte** [comment]
Quick thought, clippy nowadays takes `rust-version` into account for many lints. Does the compiler do the same, in particular for the lint for unreachable match arms? Would this be an easy addition, if it's not a thing yet?

**Comment by Nadrieril** [comment]
I didn't write it explicitly but arrays and tuples can also be empty, in the expected way. I'll add it to the OP

**Comment by RalfJung** [comment]
For arrays I hope that `[!; 0]` is considered *non-empty* as it should? Please make sure we have a test for that.

**Comment by Nadrieril** [comment]
To my own shock I did not find a test for that :') There's one now.

File: tests/ui/codemap_tests/huge_multispan_highlight.svg
Comment (Nadrieril) [reply tests/ui/codemap_tests/huge_multispan_highlight.svg:1]: Rebasing doesn't seem to be enough

File: tests/mir-opt/unreachable.as_match.UnreachablePropagation.panic-unwind.diff
Comment (Nadrieril) [reply tests/mir-opt/unreachable.as_match.UnreachablePropagation.panic-unwind.diff:35]: well spotted, ty

**Comment by Nadrieril** [comment]
@rustbot concern type-inference

I stumbled upon https://github.com/rust-lang/rust/pull/108993 which changes how we compute inhabitedness. If I follow correctly, this would mean we don't look into inferred types to determine inhabitedness. So the following, which works today, would error instead:
```rust
#![feature(min_exhaustive_patterns)]
#![allow(unused)]

fn infer<F: FnOnce() -> R, R>(_: Result<bool, R>) {}

enum Void {}
fn with_void() {
    let mut x = Ok(true);
    match x {
        Ok(_) => {}
    }
    infer::<fn() -> Void, _>(x)
}
```

Hence https://github.com/rust-lang/rust/pull/108993 becomes a breaking change. We thus need to decide whether to accept https://github.com/rust-lang/rust/pull/108993 before we stabilize `min_exhaustive_patterns`.

**Comment by cjgillot** [comment]
I'd reason the other way around: stabilize this PR and adapt https://github.com/rust-lang/rust/pull/108993 on those grounds. https://github.com/rust-lang/rust/pull/108993 is a bit too eager to declare some code unreachable, which is an issue in itself.

**Comment by Nadrieril** [comment]
Would it make sense to put https://github.com/rust-lang/rust/pull/108993 under this feature gate?

**Comment by traviscross** [comment]
It's interesting that we now have a *cycle*.  #108993 was blocked on the stabilization of exhaustive patterns.  But to stabilize exhaustive patterns, we need to land #108993 or decide not to do so (*completing the cycle...*).

**Comment by Nadrieril** [comment]
Alright, having chatted with @cjgillot and @WaffleLapkin I no longer think it would make sense for exhaustiveness to infer less than it does today. Normalization is important to patterns behaving consistently. My concern is resolved and we don't need to wait for https://github.com/rust-lang/rust/pull/108993.

**Comment by Nadrieril** [comment]
> @Nadrieril I'd like to confirm two things...
> 
>     1. I presume there is a test for matching on `pub struct Foo { x: ! }` (where the field `x` is private at the point of match) and showing that it is not considered uninhabited?

Yep! See https://github.com/Nadrieril/rust/blob/stabilize-min-exh-pats2/tests/ui/pattern/usefulness/uninhabited.rs for various tests around this.
>
>     2. You wrote that "structs with a visible uninhabited field" are considered uninhabited _unless_ they are marked with `#[non_exhaustive]` -- what does non-exhaustive have to do with this? In particular, I think of "non-exhaustive" as meaning "may add more fields" in the future, but even if that happens, there would still be an uninhabited field, so the struct is still not instantiable -- what am I missing?

It's a historical accident, I agree that `non_exhaustive` shouldn't affect struct inhabitedness. I hadn't noticed when submitting the stabilization PR, hence the confusion.

If I follow, the current FCP is approving the currently implemented (and incorrect) behaviour. I can make another PR to correct the behavior later; it's not a breaking change.

**Comment by clarfonthey** [comment]
I would assume that `#[non_exhaustive]` wouldn't affect struct inhabitedness but it would affect enum inhabitedness, since adding a variant to an uninhabited enum could make it inhabited.

Of course, this shouldn't affect dead code elimination, just matching requirements, which is what we're discussing anyway.

**Comment by nikomatsakis** [comment]
Agree that we can correct the min exhaustive pattern behavior for structs later. It'd be useful to document clearly that non exhaustive on structs implies new fields may be added in the future (including possibly public ones). Or maybe it's already documented. 

On Thu, May 2, 2024, at 9:15 PM, Guillaume Boisseau wrote:
> 
> 
>> @Nadrieril <https://github.com/Nadrieril> I'd like to confirm two things...
>> 
>> `1. I presume there is a test for matching on `pub struct Foo { x: ! }` (where the field `x` is private at the point of match) and showing that it is not considered uninhabited?
`
> Yep! See https://github.com/Nadrieril/rust/blob/stabilize-min-exh-pats2/tests/ui/pattern/usefulness/uninhabited.rs for various tests around this.
> 
>> `2. You wrote that "structs with a visible uninhabited field" are considered uninhabited _unless_ they are marked with `#[non_exhaustive]` -- what does non-exhaustive have to do with this? In particular, I think of "non-exhaustive" as meaning "may add more fields" in the future, but even if that happens, there would still be an uninhabited field, so the struct is still not instantiable -- what am I missing?
`

**Comment by Nadrieril** [comment]
I'd prefer to wait for the lint improvements [suggested by @fee1-dead](https://github.com/rust-lang/rust/pull/122792#discussion_r1533297930) before merging. I'll get to them when I find a moment.

**Comment by rs-sac** [comment]
@Nadrieril Do you think you'll be able to make progress on this feature soon?  I'm currently cleaning up a bunch of unstable feature usage, and being able to use this, if stabilized, would be quite helpful.

**Comment by Nadrieril** [comment]
Hey, thanks for the ping! I did get side-tracked by other pattern-related things. I'll allocate some time to this.

File: compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs
Comment (bjorn3) [inline compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs:591]: Does the MIR for this function keep both matches? If not, this wouldn't test what it is intended to test anymore.

Comment (Nadrieril) [reply compiler/rustc_codegen_cranelift/example/mini_core_hello_world.rs:591]: Yes: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b88ecbaaeea020ab1055a977170749a0. MIR lowering is affected by this, but only in the case of an enum with one non-empty variant and a bunch of empty variants.

**Comment by Nadrieril** [comment]
@fee1-dead I've improved the lints to my satisfaction (https://github.com/rust-lang/rust/pull/128033, https://github.com/rust-lang/rust/pull/128034). The PR has had a few rebases since you reviewed it last, would you mind having a last look over it?

**Review by fee1-dead — APPROVED** [review]

**Comment by matthiaskrgr** [comment]
@bors r-
guess this failed here? https://github.com/rust-lang/rust/pull/128719#issuecomment-2270535371

**Comment by tgross35** [comment]
@bors r-

Bors put this back into the queue after the resync

**Comment by Nadrieril** [comment]
Do you know how I could run all these tests myself?

**Comment by tgross35** [comment]
Add `try-job: dist-aarch64-apple` (can do multiple if there are other  jobs likely to fail) then just bors try

**Comment by Nadrieril** [comment]
Thank you!

**Comment by Nadrieril** [comment]
(nomicon fix is ready: https://github.com/rust-lang/nomicon/pull/445)

**Comment by Nadrieril** [comment]
> Quick thought, clippy nowadays takes `rust-version` into account for many lints. Does the compiler do the same, in particular for the lint for unreachable match arms? Would this be an easy addition, if it's not a thing yet?

@jplatte apologies, it seems I entirely overlooked your comment at the time. The compiler doesn't do the same, I doubt it is even aware of `rust-version`. I imagine you'd want this to prevent churn on unsuspecting users? The mechanism for that is editions, I don't think the churn here is sufficient to warrant that. I'll make the lint rustfixable to make this easier too.

**Comment by jplatte** [comment]
> I imagine you'd want this to prevent churn on unsuspecting users?

Not so much, actually. I think churn is much more likely to happen for libraries that don't declare `rust-version` at all, whose authors don't really think about the concept.

It's more that I think people who *do* declare a `rust-version` and want to keep it low-ish are going to find this lint annoying, because they wouldn't want to bump MSRV to satisfy it, so all those people will have to explicitly silence it (which is not the case for similar clippy lints).

> The mechanism for that is editions, I don't think the churn here is sufficient to warrant that.

What would be the harm in making it allow-by-default in current editions though? It's not that long until the next edition anyways. I think this lint will cause churn for some library writers (in the form of CI breaking, because warnings are set to deny), and I don't think the benefits of slightly shorter code is actually worth it (even though I'm actually pretty excited about this feature shipping!).

**Comment by Nadrieril** [comment]
> It's more that I think people who _do_ declare a `rust-version` and want to keep it low-ish are going to find this lint annoying, because they wouldn't want to bump MSRV to satisfy it, so all those people will have to explicitly silence it (which is not the case for similar clippy lints).

Oh yeah, I didn't think of that.

> What would be the harm in making it allow-by-default in current editions though?

None indeed. The feature is available to users who want it even without the warning. I'm not sure how to call the shots here though.

**Comment by seanmonstar** [comment]
Just to chime in here, I noticed in hyper's CI that this started failing (yes, I deny warnings in CI, it's the best way I notice changes). What's interesting is that the warning seems to be emitted in the same version that it even became possible to omit the pattern in the first place. So, in order to support older rust version (even just current stable, 1.80), I can't appease the warning. I can only turn it off.

Would it make sense to promote the lint to a warning a few releases later?

For the record, I have code that's matching on a `Result<T, Infallible>`, which an `Err(never) => match never {}` arm.

**Comment by kpreid** [comment]
I filed issue #129031 to give this matter more visibility.

**Comment by rylev** [comment]
@Nadrieril @fee1-dead seems this has a negative impact on the performance of coherence checking which I imagined is expected, right? I'm not sure if there's anything we would want to do given that there's simply more work happening. Thoughts?

**Comment by Nadrieril** [comment]
@rylev yeah, at the very least it's calculating inhabitedness of essentially all types involed in patterns, which has unavoidable cost.

**Comment by Dr-Emann** [comment]
I think I found either a bug, or a confusing error: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=08d6bead0bfb94e25a784eb88f8d059f

```rust
let x: Result<(), Infallible> = Ok(());
match *&x {
  Ok(()) => { println!("Must be OK"); }
}
```

Errors (it doesn't seem to consider `*&x` to be matching by value):
```
note: the matched value is of type `Result<(), Infallible>`
note: `Infallible` is uninhabited but is not being matched by value, so a wildcard `_` is required
```

As far as I know, I would consider that to be matching by value, but it's possible there's some "place vs value" theory that means it's not _technically_ by value.

**Comment by Nadrieril** [comment]
Indeed this isn't by-value because the `Infallible` value isn't copied: this is matching on the `*&x` place which is behind a reference. It would be matching by-value if you forced a copy with `match {*&x} { ... }`. Tbf the line above says "the matched value is of type" when it would be more accurate to say "the matched place is of type".

**Pull request merged**
